pkg/integrations/prometheus/utils/operatorutils.go

```diff
@@ -0,0 +1,243 @@
+	case utils.LetSubtract:
+		val := valueLHS - valueRHS
+		finalResult[groupID][timestamp] = val
+	case utils.LetModulo:
+		finalResult[groupID][timestamp] = math.Mod(valueLHS, valueRHS)
+	case utils.LetPower:
+		finalResult[groupID][timestamp] = math.Pow(valueLHS, valueRHS)
+	case utils.LetGreaterThan:
+		isGtr := valueLHS > valueRHS
+		if isGtr {
+			if queryOp.ReturnBool {
+				finalResult[groupID][timestamp] = 1
+				return
+			}
+			if swapped {
+				finalResult[groupID][timestamp] = valueRHS
+			} else {
+				finalResult[groupID][timestamp] = valueLHS
+			}
+		}
+	case utils.LetGreaterThanOrEqualTo:
+		isGte := valueLHS >= valueRHS
+		if isGte {
+			if queryOp.ReturnBool {
+				finalResult[groupID][timestamp] = 1
+				return
+			}
+			if swapped {
+				finalResult[groupID][timestamp] = valueRHS
+			} else {
+				finalResult[groupID][timestamp] = valueLHS
+			}
+		}
+	case utils.LetLessThan:
+		isLss := valueLHS < valueRHS
+		if isLss {
+			if queryOp.ReturnBool {
+				finalResult[groupID][timestamp] = 1
+				return
+			}
+			if swapped {
+				finalResult[groupID][timestamp] = valueRHS
+			} else {
+				finalResult[groupID][timestamp] = valueLHS
+			}
+		}
+	case utils.LetLessThanOrEqualTo:
+		isLte := valueLHS <= valueRHS
+		if isLte {
+			if queryOp.ReturnBool {
+				finalResult[groupID][timestamp] = 1
+				return
+			}
+			if swapped {
+				finalResult[groupID][timestamp] = valueRHS
+			} else {
+				finalResult[groupID][timestamp] = valueLHS
+			}
+		}
+	case utils.LetEquals:
+		if valueLHS == valueRHS {
+			if queryOp.ReturnBool {
+				finalResult[groupID][timestamp] = 1
+			} else {
+				finalResult[groupID][timestamp] = valueLHS
+			}
+		}
+	case utils.LetNotEquals:
+		if valueLHS != valueRHS {
+			if queryOp.ReturnBool {
+				finalResult[groupID][timestamp] = 1
+			} else {
+				finalResult[groupID][timestamp] = valueLHS
+			}
+		}
+	default:
+		log.Errorf("SetFinalResult: does not support using this operator: %v", queryOp.Operation)
+	}
+
+	// Logical ops can only been used between 2 vectors
+	if !queryOp.ConstantOp && IsLogicalOperator(queryOp.Operation) {
+		switch queryOp.Operation {
+		case utils.LetAnd:
+			finalResult[groupID][timestamp] = valueLHS
+		case utils.LetOr:
+			finalResult[groupID][timestamp] = valueLHS
+		case utils.LetUnless:
+			finalResult[groupID][timestamp] = valueLHS
+		default:
+			log.Errorf("SetFinalResult: does not support using this operator: %v", queryOp.Operation)
+		}
+	}
+}
+
+// This method will return uniuqe tag key value pair combination str, we refer it as labelStr.
+// If matchingLabels = [tag1], groupIDStr = "metricName{tag1:val1,tag2:val2,tag3:val3}""
+// When includeColumns is true, extract labels and generate labelStr from matchingLabels based on the previous groupIDStr. The result is {tag1:val1}
+// Otherwise, exclude the labels, so the result is {tag2:val2,tag3:val3}
+func ExtractMatchingLabelSet(groupIDStr string, matchingLabels []string, includeColumns bool) string {
```

Comment:
I like that this comment describes what is supposed to happen pretty clearly. But I don't think this is what the function will actually do. You should add some unit tests. I'm pretty sure that instead of returning something like `" {tag2:val2,tag3:val3}"`, it would return something like `"{tag2:val2,tag3:val3,"`

Reply:
Yeah, i will update the comments to make it more accurate